python/ray/dag/compiled_dag_node.py

```diff
@@ -1539,6 +1499,124 @@ def _get_node_id(self, actor_handle: Optional["ray.actor.ActorHandle"]) -> str:
+            raise ValueError(
+                "Outputs cannot be transferred via NCCL because the driver "
+                "cannot participate in the NCCL group"
+            )
+        elif not isinstance(node, ClassMethodNode):
+            raise ValueError(
+                "NCCL P2P operation is only supported with ClassMethodNode"
+            )
+
+        send_actor_handle: "ray.actor.ActorHandle" = node._get_actor_handle()
+        assert send_actor_handle is not None
+        send_node = P2PSendNode(
+            method_args=(node,),
+            other_args_to_resolve={
+                PARENT_CLASS_NODE_KEY: send_actor_handle,
+                BIND_INDEX_KEY: send_actor_handle._ray_dag_bind_index,
+            },
+        )
+        send_actor_handle._ray_dag_bind_index += 1
+        send_node._type_hint = node.type_hint
+        send_node._original_type_hint = node._original_type_hint
+        node._type_hint = ChannelOutputType()
+        node._original_type_hint = None
+        node_to_p2p_send_node[node] = send_node
+        self._add_node(send_node)
+
+    def _add_p2p_recv_nodes(
+        self,
+        node: "ray.dag.DAGNode",
+        node_to_p2p_send_node: Dict["ray.dag.DAGNode", "ray.dag.P2PSendNode"],
+    ) -> None:
+        """
+        Add a NCCL P2P recv node to the DAG for each upstream node that requires
+        NCCL send.
+
+        Args:
+            node: A DAG node.
+            p2p_send_dag_nodes: The set of DAG nodes that require NCCL send.
+            node_to_p2p_send_node: A dictionary mapping DAG nodes to their
+                corresponding NCCL P2P send nodes.
+        """
+        from ray.dag import ClassMethodNode, DAGNode, MultiOutputNode
+        from ray.dag.communication_node import P2PRecvNode
+
+        new_args = []
+        for arg in node.get_args():
+            if not isinstance(arg, DAGNode) or arg not in node_to_p2p_send_node:
+                new_args.append(arg)
+                continue
+
+            if isinstance(node, MultiOutputNode):
+                raise ValueError(
+                    "Outputs cannot be transferred via NCCL because the driver "
+                    "cannot participate in the NCCL group"
+                )
+            elif not isinstance(node, ClassMethodNode):
+                raise ValueError(
+                    "NCCL P2P operation is only supported with ClassMethodNode"
+                )
+
+            send_node = node_to_p2p_send_node[arg]
+            recv_actor_handle: "ray.actor.ActorHandle" = node._get_actor_handle()
+            assert recv_actor_handle is not None
+            recv_node = P2PRecvNode(
+                method_args=(send_node,),
+                other_args_to_resolve={
+                    PARENT_CLASS_NODE_KEY: recv_actor_handle,
+                    BIND_INDEX_KEY: recv_actor_handle._ray_dag_bind_index,
+                },
+            )
+            recv_actor_handle._ray_dag_bind_index += 1
```

Comment:
Hmm I don't understand how this fix works. Couldn't this give the send/recv task a higher bind index than a task that depends on it?

Reply:
It would be better to assign the bind indices in a way that doesn't modify any state in-place.

Reply:
This bind index is at the level of `DAGNode`. After we create the P2P send/recv nodes, we call `CompiledDAG._add_node()`, which assigning increasing task index at the level of `CompiledDAG`.

Reply:
I think these new nodes can be assigned any bind indicies. If following the tradition of increasing actor handle bind index, I think we need to modify the actor state.

Reply:
To illustrate this, now I set their bind indices to None when creating send/recv nodes. They use a value of -1 when sorting executable tasks.